src/models/action.rs

```diff
@@ -0,0 +1,93 @@
+use chrono::NaiveDateTime;
+use diesel::prelude::*;
+use diesel::{
+    deserialize::{self, FromSql},
+    pg::Pg,
+    serialize::{self, Output, ToSql},
+    sql_types::Integer,
+};
+use std::io::Write;
+
+use crate::models::{ApiToken, User, Version};
+use crate::schema::*;
+
+#[derive(Debug, Clone, Copy, FromSqlRow, AsExpression)]
+#[repr(u32)]
+pub enum VersionAction {
+    Publish = 0,
+    Yank = 1,
+    Unyank = 2,
+}
+
+impl Into<i32> for VersionAction {
+    fn into(self) -> i32 {
+        match self {
+            VersionAction::Publish => 0,
+            VersionAction::Yank => 1,
+            VersionAction::Unyank => 2,
+        }
+    }
+}
+
+impl FromSql<Integer, Pg> for VersionAction {
+    fn from_sql(bytes: Option<&[u8]>) -> deserialize::Result<Self> {
+        match <i32 as FromSql<Integer, Pg>>::from_sql(bytes)? {
+            0 => Ok(VersionAction::Publish),
+            1 => Ok(VersionAction::Yank),
+            2 => Ok(VersionAction::Unyank),
+            n => Err(format!("unknown version action: {}", n).into()),
+        }
+    }
+}
+
+impl ToSql<Integer, Pg> for VersionAction {
+    fn to_sql<W: Write>(&self, out: &mut Output<'_, W, Pg>) -> serialize::Result {
+        let value: i32 = (*self).into();
+        ToSql::<Integer, Pg>::to_sql(&value, out)
+    }
+}
+
+#[derive(Debug, Clone, Copy, Queryable, Identifiable, Associations)]
+#[belongs_to(Version)]
+#[belongs_to(User, foreign_key = "owner_id")]
+#[belongs_to(ApiToken, foreign_key = "owner_token_id")]
+#[table_name = "version_owner_actions"]
+pub struct VersionOwnerAction {
+    pub id: i32,
+    pub version_id: i32,
+    pub owner_id: i32,
+    pub owner_token_id: Option<i32>,
+    pub action: VersionAction,
+    pub time: NaiveDateTime,
+}
+
+#[derive(Copy, Clone, Debug, Insertable)]
+#[table_name = "version_owner_actions"]
+pub struct NewVersionOwnerAction {
+    pub version_id: i32,
+    pub owner_id: i32,
+    pub owner_token_id: Option<i32>,
+    pub action: i32,
+}
+
+impl NewVersionOwnerAction {
+    pub fn new(
+        version_id: i32,
+        owner_id: i32,
+        owner_token_id: Option<i32>,
```

Comment:
This is None if it's session auth. Is that ok? Is there a better way to represent that?